addon/chart-data-updater.js

```diff
@@ -37,20 +37,36 @@ export default Ember.Object.extend({
     var chart = this.get('chart');
     var needUpdate = false;
 
-    data.forEach(function(segment, i) {
-      if (typeof chart.segments[i] !== 'undefined') {
+    data.forEach(function(segment) {
+      var currentSegment = chart.segments.findBy('label', segment.label);
+      if(chart.segments.findBy('label', segment.label)) {
         segment.value = segment.value || 0;
-        if (chart.segments[i].value !== segment.value) {
-          chart.segments[i].value = segment.value;
+        if (currentSegment.value !== segment.value) {
```

Comment:
what happens if we set the valuable regardless? Does calling update when the value doesn't change do anything? We might be able to remove a lot of code if assigning the value anyway is feasible.